init.lua

```diff
@@ -1,77 +1,411 @@
+	if timer < update_interval then
+		return
+	end
+	timer = 0
+
+	-- Run all custom player callbacks for each player
+	for _, player in pairs(minetest.get_connected_players()) do
+		for _,callback in pairs(update_player_callbacks) do
+			callback(player)
+		end
+	end
+
+	-- Run all custom callbacks
+	for _,callback in pairs(update_callbacks) do
+		callback()
+	end
+
+	-- Look at each tracked entity and update its position
+	for uid, entity in pairs(tracked_entities) do
+		if is_entity_valid(entity) then
+			update_entity(entity)
+		else
+			-- If the entity no longer exists, stop tracking it
+			tracked_entities[uid] = nil
+		end
+	end
+
+	-- Recalculate light levels
+	for pos,_ in pairs(light_recalcs) do
+		recalc_light(pos)
+	end
+	light_recalcs = {}
+end
+
+-- Recalculate the total light level for a given position and update the light level there
+function recalc_light(pos)
+	-- If not in active lights list we can't do anything
+	if not active_lights[pos] then return end
+
+	-- Calculate the light level of the node
+	local any_light = false
+	local max_light = 0
+	for id, light_level in pairs(active_lights[pos]) do
+		any_light = true
+		if light_level > max_light then
+			max_light = light_level
+		end
+	end
+
+	-- Convert the position back to a vector
+	local pos_vec = minetest.string_to_pos(pos)
+
+	-- If no items in this position, delete it from the list and remove any light node
+	if not any_light then
+		active_lights[pos] = nil
+		minetest.swap_node(pos_vec, { name = "air" })
 		return
 	end
 
-	light_level = light_level or ((itemdef.light_source or 0) - level_delta)
+	-- If no light in this position remove any light node
+	if max_light == 0 then
+		minetest.swap_node(pos_vec, { name = "air" })
+		return
+	end
+
+	-- Limit the light level
+	max_light = math.min(max_light, minetest.LIGHT_MAX)
+
+	-- Get the current light level in this position
+	local name = minetest.get_node(pos_vec).name
+	local old_value = wielded_light.level_of_wielded_light(name) or 0
+
+	-- If the light level has changed, set the coresponding light node and initiate the cleanup timer
+	if old_value ~= max_light then
+		minetest.swap_node(pos_vec, {
+			name = wielded_light.wielded_light_of_level(max_light)
+		})
+		minetest.get_node_timer(pos_vec):start(cleanup_interval)
+	end
+end
+
+-- Add light to active light list and mark position for update
+function add_light(pos, id, light_level)
+	if not active_lights[pos] then
+		active_lights[pos] = {}
+	end
+	if active_lights[pos][id] ~= light_level then
+		-- minetest.log("error", "add "..id.." "..pos.." "..tostring(light_level))
+		active_lights[pos][id] = light_level
+		light_recalcs[pos] = true
+	end
+end
+
+-- Remove light from active light list and mark position for update
+function remove_light(pos, id)
+	if not active_lights[pos] then return end
+	-- minetest.log("error", "rem "..id.." "..pos)
+	-- minetest.after(update_interval, function ()
+	active_lights[pos][id] = nil
+	light_recalcs[pos] = true
+	-- end)
```

Comment:
@SmallJoker

> Sometimes entire mapblocks flicker while moving.

Thanks for testing, I thought I might be the only one seeing that. Uncommneting this delay fixes it for me.
The two lists are decoupled enough that delaying it should have no side effects, I just didn't want to leave this in without knowing it was an issue as it seems a little more random than other know causes of flickering.
Basically, the light doesn't seem to update immediately, leaving the last light in for another round seems to smooth that out - some flickering still noticeable around the edges if you look closely, but I think that's unavoidable.

I'll fix this now.

Reply:
Actually I just thought of  way to break this, let me think about it for a sec

Reply:
Alright give that a shot, pretty sure it can't break and fixes the flickering

Reply:
How could this cause bugs? Is the light information missing in the timer/globalstep whenever the position changes?

Let me know when you're done; and I'll give it a try again.

Reply:
It's done, that's what I meant with the last comment
> Alright give that a shot, pretty sure it can't break and fixes the flickering

So, originally I had
```
minetest.after(update_interval, function ()
	active_lights[pos][id] = nil
	light_recalcs[pos] = true
end)
```
which I noticed may cause an issue if I move away from and then back to the same spot within one update_interval. When I get back to the spot it may erroneously remove the light source as it had been marked for removal in the future.

The code now contains (something like) this:
```
active_lights[pos][id] = nil
minetest.after(update_interval, function ()
	light_recalcs[pos] = true
end)
```

Which fixes it in pretty much the same way. It updates the list synchronously, but doesn't ask for a synchronous update, it ask for it on the next update step. This means that the list will always be correct, but the light in the old position will hang around for another step before the position is updated.